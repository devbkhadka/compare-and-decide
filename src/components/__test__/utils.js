import React from 'react'
import { Provider } from 'react-redux'
import { IntlProvider } from 'react-intl'
import { render } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureMockStore(middlewares)

export function renderWithIntl(component) {
    return render(<IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>)
}

export function renderWithProvider(component, initialState={}) {
    const store = mockStore(initialState)
    return [render(<Provider store={store}>
        <IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>
    </Provider>), store]
}