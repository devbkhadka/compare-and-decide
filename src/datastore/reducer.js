import { ADD_ITEM_WITH_TITLE } from './constants'

const initialState = {
    items:[]
}

export default function (state=initialState, payload) {
    switch(payload.type) {
        case ADD_ITEM_WITH_TITLE:
            if(payload.data){
                return {...state, items: [...state.items, payload.data]}
            }
        
        default:
            return state
    }
}