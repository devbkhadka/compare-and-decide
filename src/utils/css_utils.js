export function joinClss(...classes){
    return classes.join(' ')
}