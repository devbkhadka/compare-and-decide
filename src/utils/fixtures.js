const dummyItems = JSON.stringify({
    Item1: {
        title: 'Item1',
        values: {}
    },
    Item2: {
        title: 'Item2',
        values: {}
    },
    Item3: {
        title: 'Item3',
        values: {}
    },
    Item4: {
        title: 'Item4',
        values: {}
    }
})

const dummyAttributes = JSON.stringify(['attr1', 'attr2', 'attr3', 'attr4'])

const dummyItemsWithValues = JSON.stringify({
    Item1: {
        title: 'Item1',
        values: {
            'attr1': 'Item1-Value1',
            'attr2': 'Item1-Value2',
            'attr3': 'Item1-Value3',
            'attr4': 'Item1-Value4'
        }
    },
    Item2: {
        title: 'Item2',
        values: {
            'attr2': 'Item2-Value2',
            'attr4': 'Item4-Value4'
        }
    },
    Item3: {
        title: 'Item3',
        values: {
            'attr1': 'Item3-Value1',
            'attr2': 'Item3-Value2',
            'attr3': 'Item3-Value3',
        }
    },
    Item4: {
        title: 'Item4',
        values: {}
    }
})

const dummyAttributesDelAttr3 = JSON.stringify(['attr1', 'attr2', 'attr4'])
const dummyItemsDelAttr3 = JSON.stringify({
    Item1: {
        title: 'Item1',
        values: {
            'attr1': 'Item1-Value1',
            'attr2': 'Item1-Value2',
            'attr4': 'Item1-Value4'
        }
    },
    Item2: {
        title: 'Item2',
        values: {
            'attr2': 'Item2-Value2',
            'attr4': 'Item4-Value4'
        }
    },
    Item3: {
        title: 'Item3',
        values: {
            'attr1': 'Item3-Value1',
            'attr2': 'Item3-Value2',
        }
    },
    Item4: {
        title: 'Item4',
        values: {}
    }
})

export const getDummyItems = ()=> JSON.parse(dummyItems)
export const getDummyItemsWithValues = ()=> JSON.parse(dummyItemsWithValues)
export const getDummyItemsDelAttr3 = ()=> JSON.parse(dummyItemsDelAttr3)
export const getDummyAttributes = () => JSON.parse(dummyAttributes)
export const getDummyAttributesDelAttr3 = () => JSON.parse(dummyAttributesDelAttr3)
export const getDummyState = ()=> ({
    items: getDummyItems(),
    attributes: getDummyAttributes(),
})