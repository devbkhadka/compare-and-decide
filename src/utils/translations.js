import React from 'react'
import { FormattedMessage } from 'react-intl'

export default function formattedText(message) {
    let key
    for(key in message) {break}
    return <FormattedMessage {...message[key]} />
}