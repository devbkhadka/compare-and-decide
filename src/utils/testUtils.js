import React from 'react'
import { Provider } from 'react-redux'
import { IntlProvider } from 'react-intl'
import { render } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'
import {middleWares} from '../datastore/store'
import ReduxThunk from 'redux-thunk'


const mockStore = configureMockStore(middleWares)

export function renderWithIntl(component) {
    return render(<IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>)
}

export function renderWithProvider(component, initialState={language: 'en'}) {
    const store = mockStore(initialState)
    return [render(<Provider store={store}>
        <IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>
    </Provider>), store]
}