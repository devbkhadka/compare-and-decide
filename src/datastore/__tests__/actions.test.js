import {addItemWithTitle, 
    updateLanguage, 
    addAttribute,
    updateItems,
    loadStateFromStorage,
    ITEMS_KEY,
    ATTRIBUTES_KEY,
    updateAttributes,
    updateItemAttributeValue,
    setPreferedItemTitle,
    PREFETED_ITEM_TITLE_KEY,
    updatePreferedItemTitle as updatePreferedItemTitle,
    rejectItem,
    deleteAttribute,
    deleteItem,
    clearComparisionStatus
} from '../actions'
import { COMPARISION_STATUS } from '../reducer'
import * as fixtures from '../../utils/fixtures'

function assertItemsUpdatedAndSaved(items, dispatch) {
    expect(localStorage.setItem).toBeCalledWith(ITEMS_KEY, JSON.stringify(items))
    expect(dispatch).toBeCalledWith(updateItems(items))
}

function assertItemsLoadedFromStorage(items, dispatch){
    expect(localStorage.getItem).toBeCalledWith(ITEMS_KEY)
    expect(dispatch).toBeCalledWith(updateItems(items))
}

function assertAttributesUpdatedAndSaved(expectedAttributes, mockDispatch){
    expect(localStorage.setItem).toBeCalledWith(ATTRIBUTES_KEY, JSON.stringify(expectedAttributes))
    expect(mockDispatch).toBeCalledWith(updateAttributes(expectedAttributes))
}

function assertAttributeLoadedFromStorage(attributes, dispatch) {
    expect(localStorage.getItem).toBeCalledWith(ATTRIBUTES_KEY)
    expect(dispatch).toBeCalledWith(updateAttributes(attributes))
}

function assertPreferedItemUpdatedAndSaved(itemTitle, mockDispatch) {
    expect(localStorage.setItem).toBeCalledWith(PREFETED_ITEM_TITLE_KEY, itemTitle)
    expect(mockDispatch).toBeCalledWith(updatePreferedItemTitle(itemTitle))
}

function assertPreferedItemLoadedFromStorage(itemTitle, mockDispatch) {
    expect(localStorage.getItem).toBeCalledWith(PREFETED_ITEM_TITLE_KEY)
    expect(mockDispatch).toBeCalledWith(updatePreferedItemTitle(itemTitle))
}

describe('Test actions', ()=>{
    jest.spyOn(Storage.prototype, 'setItem').mockImplementation(()=>null)
    jest.spyOn(Storage.prototype, 'getItem')

    beforeEach(()=>{
        Storage.prototype.setItem.mockClear()
        Storage.prototype.getItem.mockClear()
    })

    describe('Items', ()=>{
        test('addItemWithTitle add item and save it to localstorage', ()=>{
            const items = fixtures.getDummyItems()
            const Item4 = items.Item4
            delete items.Item4

            const mockDispatch = jest.fn()
            addItemWithTitle(Item4.title)(mockDispatch, ()=>({items}))

            const expectedItems = fixtures.getDummyItems()
            assertItemsUpdatedAndSaved(expectedItems, mockDispatch)

        })

        test('addItemWithTitle will not add new item if it already exists', ()=>{
            const dispatch = jest.fn()
            addItemWithTitle('Item 1')(dispatch, ()=>({items:{['Item 1']: {}}}))

            expect(localStorage.setItem).not.toBeCalled()
            expect(dispatch).not.toBeCalled()
        })

        test('updateItem returns correct item object', ()=>{
            const items = fixtures.getDummyItems()
            expect(updateItems(items)).toMatchSnapshot()
        })

        test('loadStateFromStorage loads items from local storage', ()=>{
            const mockItems = fixtures.getDummyItems()
            const mockDispatch = jest.fn()
            Storage.prototype.getItem.
                mockImplementation((key)=>key==ITEMS_KEY? JSON.stringify(mockItems): null)
            
            loadStateFromStorage()(mockDispatch)

            assertItemsLoadedFromStorage(mockItems, mockDispatch)
        })
        
        it('doesn\'t update state if localstorage donot have items ', ()=>{
            const mockDispatch = jest.fn()
            Storage.prototype.getItem.
                mockImplementation(()=>null)

            loadStateFromStorage()(mockDispatch)
            
            expect(mockDispatch).not.toBeCalled()
        })
    })

    describe('Item Deletion', ()=>{
        it('should delete item and save the items to localstorage', ()=>{
            const items = fixtures.getDummyItems()
            const mockDispatch = jest.fn()
            deleteItem('Item2')(mockDispatch, ()=>({items}))
            const expectedItems = fixtures.getDummyItems()
            delete expectedItems.Item2
            assertItemsUpdatedAndSaved(expectedItems, mockDispatch)
        })
    })

    describe('Attributes', ()=>{
        test('addAttribute adds atributes to state and save it to local storage', ()=>{
            const attributes = fixtures.getDummyAttributes()
            const mockDispatch = jest.fn()

            const expectedAttributes = [...attributes, 'attr5']
            addAttribute('attr5')(mockDispatch, ()=>({attributes}))
            
            assertAttributesUpdatedAndSaved(expectedAttributes, mockDispatch)
        })

        test('addAttribute doesn\'t add empty attribute', ()=>{
            const attributes = fixtures.getDummyAttributes()
            const mockDispatch = jest.fn()

            addAttribute('')(mockDispatch, ()=>({attributes}))
            expect(mockDispatch).not.toBeCalled()
        })

        test('addAttribute doesn\'t add duplicate attribute', ()=>{
            const attributes = fixtures.getDummyAttributes()
            const mockDispatch = jest.fn()

            addAttribute('attr1')(mockDispatch, ()=>({attributes}))
            expect(mockDispatch).not.toBeCalled()
        })

        test('updateAttributes returns action correctly', ()=>{
            expect(updateAttributes(fixtures.getDummyAttributes())).toMatchSnapshot()
        })

        test('loadStateFromStorage loads attributes from storage', ()=>{
            const attributes = fixtures.getDummyAttributes()
            Storage.prototype.getItem.
                mockImplementation((key)=>key==ATTRIBUTES_KEY? JSON.stringify(attributes): null)
            const mockDispatch = jest.fn()
            
            loadStateFromStorage()(mockDispatch)
            assertAttributeLoadedFromStorage(attributes, mockDispatch)

        })

        test('loadStateFromStorage donot update attributes if not exists on storage', ()=>{
            const mockDispatch = jest.fn()
            Storage.prototype.getItem.
                mockImplementation((key)=>null)

            loadStateFromStorage()(mockDispatch)
            expect(mockDispatch).not.toBeCalled()
        })

        test('deleteAttribute deletes given attribute and save attributes to storage', ()=>{
            const mockDispatch = jest.fn()
            const attributes = fixtures.getDummyAttributes()
            
            deleteAttribute(attributes[2])(mockDispatch, ()=>({attributes}))

            const expectedAttributes = fixtures.getDummyAttributesDelAttr3()
            assertAttributesUpdatedAndSaved(expectedAttributes, mockDispatch)
        })

        test('deleteAttribute removes deleted attributes from all items', ()=>{
            const mockDispatch = jest.fn()
            const attributes = fixtures.getDummyAttributes()
            const items = fixtures.getDummyItemsWithValues()
            
            deleteAttribute(attributes[2])(mockDispatch, ()=>({attributes, items}))

            const expectedItems = fixtures.getDummyItemsDelAttr3()
            assertItemsUpdatedAndSaved(expectedItems, mockDispatch)
        })

    })

    describe('updateItemAttributeValue', ()=>{
        it('should update item attribute and save it to storage', ()=>{
            const state = fixtures.getDummyState()

            const mockDispatch = jest.fn()
            const func =updateItemAttributeValue(state.items.Item1, 'Fake Attribute', 'Fake Value')
            func(mockDispatch, ()=>state)
            
            state.items['Item1'].values['Fake Attribute'] = 'Fake Value'

            assertItemsUpdatedAndSaved(state.items, mockDispatch)
        })
    })

    describe('setPreferedItemTitle', ()=>{
        const state = fixtures.getDummyState()

        beforeEach(()=>{
            jest.resetAllMocks()
        })

        it('should set prefered item and save it to localstorage', ()=>{
            const mockDispatch = jest.fn()
            const mockGetState = ()=>state
            setPreferedItemTitle(state.items['Item2'].title)(mockDispatch, mockGetState)
            const expectedItem = state.items['Item2']
            expectedItem.comparisionStatus = COMPARISION_STATUS.ITEM_PREFERED

            assertItemsUpdatedAndSaved(state.items, mockDispatch)
            assertPreferedItemUpdatedAndSaved(expectedItem.title, mockDispatch)
        })

        it('loadStateFromStorage should load preferedItemTitle', ()=>{
            Storage.prototype.getItem.
                mockImplementation((key)=>key==PREFETED_ITEM_TITLE_KEY? 'saved prefered item title': null)
            const mockDispatch = jest.fn()
            
            loadStateFromStorage()(mockDispatch)
            assertPreferedItemLoadedFromStorage('saved prefered item title', mockDispatch)
        })
    })

    describe('Clear Comparision Status', ()=>{
        it('should clear comparision status of rejected item', ()=>{
            const items = fixtures.getDummyItems()
            items.Item2.comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED
            const mockDispatch = jest.fn()
            clearComparisionStatus(items.Item2.title)(mockDispatch, ()=>({items}))

            const expectedItems = fixtures.getDummyItems()
            assertItemsUpdatedAndSaved(expectedItems, mockDispatch)
        })
        it('should clear comparision status of prefered item', ()=>{
            const items = fixtures.getDummyItems()
            items.Item2.comparisionStatus = COMPARISION_STATUS.ITEM_PREFERED
            const mockDispatch = jest.fn()
            clearComparisionStatus(items.Item2.title)(mockDispatch, ()=>(
                {items, preferedItemTitle: items.Item2.title}))

            const expectedItems = fixtures.getDummyItems()
            assertItemsUpdatedAndSaved(expectedItems, mockDispatch)
            assertPreferedItemUpdatedAndSaved(undefined, mockDispatch)
        })
    })

    test('rejectItem sets item to be reject and save items to localstorege', ()=>{
        const state = fixtures.getDummyState()
        const mockDispatch = jest.fn()
        const item2 = state.items[Object.keys(state.items)[1]]
        rejectItem(item2.title)(mockDispatch, ()=>state)
        item2.comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED
        assertItemsUpdatedAndSaved(state.items, mockDispatch)
    })


    test('updateLanguage action correctly created', ()=> {
        expect(updateLanguage('np')).toMatchSnapshot()
    })

})
