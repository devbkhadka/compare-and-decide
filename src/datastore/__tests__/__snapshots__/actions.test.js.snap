// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test actions Attributes updateAttributes returns action correctly 1`] = `
Object {
  "data": Array [
    "attr1",
    "attr2",
    "attr3",
    "attr4",
  ],
  "type": "UPDATE_ATTRIBUTES",
}
`;

exports[`Test actions Items updateItem returns correct item object 1`] = `
Object {
  "data": Object {
    "Item1": Object {
      "title": "Item1",
      "values": Object {},
    },
    "Item2": Object {
      "title": "Item2",
      "values": Object {},
    },
    "Item3": Object {
      "title": "Item3",
      "values": Object {},
    },
    "Item4": Object {
      "title": "Item4",
      "values": Object {},
    },
  },
  "type": "UPDATE_ITEMS",
}
`;

exports[`Test actions updateLanguage action correctly created 1`] = `
Object {
  "data": "np",
  "type": "UPDATE_LANGUAGE",
}
`;
